robot/impl/local_robot_test.go

```diff
@@ -3517,3 +3517,18 @@ func TestSendTriggerConfig(t *testing.T) {
 	}
 	test.That(t, len(actualR.triggerConfig), test.ShouldEqual, 1)
 }
+
+func TestRestartModule(t *testing.T) {
+	ctx := context.Background()
+	logger := logging.NewTestLogger(t)
+	simplePath := rtestutils.BuildTempModule(t, "examples/customresources/demos/simplemodule")
+	mod := &config.Module{Name: "restartSingleModule-test", ExePath: simplePath, Type: config.ModuleTypeLocal}
+	r := setupLocalRobot(t, ctx, &config.Config{Modules: []config.Module{*mod}}, logger)
+	test.That(t, mod.LocalVersion, test.ShouldBeEmpty)
+
+	// test restart. note: we're not testing that the PID rolls over because we don't have access to
+	// that state. 'no error' + 'version incremented' is a cheap proxy for that.
+	err := r.(*localRobot).restartSingleModule(ctx, mod)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, mod.LocalVersion, test.ShouldResemble, "0.0.1")
```

Comment:
To make sure I understand the fuller lifecycle of a real viam-server here. If this test were to then:
* Create a new config object from json: `cfg, err := config.Read(ctx, cfgPath, logger)`
* Use that for a reconfigure: `r.Reconfigure(cfg)`

I expect that `cfg.Modules[0].LocalVersion` value to be the default value/empty string. And the robot would consider this a diff and restart the module? Presumably unpackaging it back into the "0.0.0" directory (or whatever gets used for the default scenario).

Reply:
the state is stored in the localRobot, not in the config

so in your example, if `r` is reused, the fresh `config.Read` won't overwrite the version state; it will be restored by the `r.applyLocalModuleVersions(newConfig)` call

in `reconfigure` at [L1168 here](https://github.com/viamrobotics/rdk/pull/4114/files#diff-0b975da6011656cd3f2caf0435023573bdb5ac7c6a425e7917dd5cc9d6be63e4R1168)

Reply:
This makes sense -- thanks for pointing me to the important missing link here!

Reply:
fwiw it would be very nice if we could store it in config; would make it way easier to understand the state for reloading + local tarball

I often start these PRs by mutating the Config struct and then am confused for an hour bc it is getting overwritten or I am mutating a copy

Reply:
Agreed, I'd like to have a better story here as well.